src/exchange_asset.rs

```diff
@@ -0,0 +1,533 @@
+    state::{accepted_subscriptions_read, asset_exchange_storage, config_read},
+};
+
+pub fn try_issue_asset_exchanges(
+    deps: DepsMut<ProvenanceQuery>,
+    info: MessageInfo,
+    asset_exchanges: Vec<IssueAssetExchange>,
+) -> ContractResponse {
+    let state = config_read(deps.storage).load()?;
+    let accepted = accepted_subscriptions_read(deps.storage)
+        .may_load()?
+        .unwrap_or_default();
+    let mut storage = asset_exchange_storage(deps.storage);
+
+    if info.sender != state.gp {
+        return contract_error("only gp can issue redemptions");
+    }
+
+    for issuance in asset_exchanges {
+        if !accepted.contains(&issuance.subscription) {
+            return contract_error("subscription not accepted");
+        }
+
+        let mut existing = storage
+            .may_load(issuance.subscription.as_bytes())?
+            .unwrap_or_default();
+
+        existing.push(issuance.exchange.clone());
+
+        storage.save(issuance.subscription.as_bytes(), &existing)?;
+    }
+
+    Ok(Response::default())
+}
+
+pub fn try_cancel_asset_exchanges(
+    deps: DepsMut<ProvenanceQuery>,
+    info: MessageInfo,
+    cancellations: Vec<IssueAssetExchange>,
+) -> ContractResponse {
+    let state = config_read(deps.storage).load()?;
+    let mut storage = asset_exchange_storage(deps.storage);
+
+    if info.sender != state.gp {
+        return contract_error("only gp can cancel redemptions");
+    }
+
+    for cancel in &cancellations {
+        let mut existing = storage
+            .may_load(cancel.subscription.as_bytes())?
+            .ok_or("no asset exchange found for subscription")?;
+
+        let index = existing
+            .iter()
+            .position(|e| &cancel.exchange == e)
+            .ok_or("no asset exchange found for subcription")?;
+        existing.remove(index);
+
+        storage.save(cancel.subscription.as_bytes(), &existing)?;
+    }
+
+    Ok(Response::default())
+}
+
+pub fn try_complete_asset_exchange(
+    deps: DepsMut<ProvenanceQuery>,
+    env: Env,
+    info: MessageInfo,
+    exchange: AssetExchange,
+    to: Option<Addr>,
+    memo: Option<String>,
+) -> ContractResponse {
+    let state = config_read(deps.storage).load()?;
+    let mut storage = asset_exchange_storage(deps.storage);
+
+    let mut existing = storage
+        .may_load(info.sender.as_bytes())?
+        .ok_or("no asset exchange found for subscription")?;
+
+    let index = existing
+        .iter()
+        .position(|e| &exchange == e)
+        .ok_or("no asset exchange found for subcription")?;
+    existing.remove(index);
+
+    storage.save(info.sender.as_bytes(), &existing)?;
+
+    if let Some(date) = &exchange.date {
+        match date {
+            ExchangeDate::Due(epoch_seconds) => {
+                if epoch_seconds < &env.block.time.seconds() {
+                    return contract_error("exchange past due");
+                }
+            }
```

Comment:
Is this logic enforcing that if an exchange is past due it can not be executed? Is that too rigid of a requirement?

Reply:
if a due date exists, asset exchanges can always be cancelled / reissued